---
title: "Small Intestine Power Analysis"
output: html_document
date: "2022-12-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r, loading, include=FALSE}
# package loading
library(RnaSeqSampleSize)
library(dplyr)
library(purrr)
library(ggplot2)
library(biomaRt)
set.seed(538734517)
```

## Parsing DataMat

```{r, file_parsing, warning=FALSE, tidy=TRUE}
# file parsing and cleaning
d <- read.delim("/home/rstudio/slee_poweranalysis/metadata/GSE143290_SI_study_RawCounts.txt", header=TRUE, sep="\t")
dataMat <- d[, c(8, 9, 10, 11, 12, 13)]
rownames(dataMat) <- d$Geneid
colnames(dataMat) <- c("VAS_Control_1", "VAS_Control_2", "VAS_Control_3", "VAS_Infected_1", "VAS_Infected_2", "VAS_Infected_3")
head(dataMat)
```

#IFNB, use string-db


## Power analysis

RNAseqsamplesize [@Zhao2018] was used to do the analysis. With expected fold change between groups = 2, FDR set = 0.01 and number of samples = 5, power was computed to be 0.086 for the selected genes of interest which is very low. Some gene IDs pulled from Biomart are not in the dataset because they are alleles on alternative sequences (i.e. rapsn and Fbxo32) Dispersion was also computed as 0.04754.

```{r, power, tidy=TRUE, warning=FALSE}
# estimate gene read count and dispersion distribution
distribution <- est_count_dispersion(dataMat,group=c(0,0,0,1,1,1))
genelist <- {}
genelist$Gene.name <- c("Trex1", "Ifrd1", "Wrn","Atrip","Exo1","Rnaseh2b","Mb21d1","Rnaseh2a","Nme1","Gzma","Rnaseh2c")

power <- est_power_distribution(n=5,f=0.01,rho=2,distributionObject=distribution,selectedGenes=c("Trex1", "Ifrd1", "Wrn","Atrip","Exo1","Rnaseh2b","Mb21d1","Rnaseh2a","Nme1","Gzma","Rnaseh2c"),storeProcess = TRUE)
mean(power$power) # 0.2187556
```

From changing around FDR (`fdr` parameter in `est_power_curve()`) and coverage (`lambda0` parameter in `est_power_curve()`) as well as making an optimization plot, it appears that at least 10 samples are needed with an average coverage > 10 reads/gene depending on desired FDR in order to achieve >80% power. To be precise, 10 samples with 10 coverage and FDR=0.05 will give 82% power. To be safe, for the genes of interest we would need at least 13 samples to achieve 80% power.

```{r, power_curve, tidy=TRUE, echo=FALSE, fig.cap="Red line is FDR=0.01, coverage=5. Blue line is FDR=0.05, coverage=5. Purple line is FDR=0.01, coverage=10. Green line is FDR=0.05, coverage=10. Yellow line is FDR=0.05, coverage=20."}
# power vs sample size plot for different coverage / FDR
coverage5_fdr1 <- est_power_curve(n=40,f=0.01,rho=2,lambda0=5,phi0=0.05714, m = 52000, m1=71)
coverage10_fdr1 <- est_power_curve(n=40,f=0.01,rho=2,lambda0=10,phi0=0.05714)
coverage5_fdr5 <- est_power_curve(n=40,f=0.05,rho=2,lambda0=5,phi=0.05714)
coverage10_fdr5 <- est_power_curve(n=40,f=0.05,rho=2,lambda0=292,phi=0.05714, m = 52000, m1=71)
coverage20_fdr5 <- est_power_curve(n=40,f=0.05,rho=2,lambda0=20,phi=0.05714)
plot_power_curve(list(coverage5_fdr1,coverage5_fdr5,coverage10_fdr1,coverage10_fdr5,coverage20_fdr5))
```

### est_power_curve

```{r, optimize, tidy=TRUE, echo=FALSE, fig.cap="Blue to brown gradient shows power from 0 to 1. Here FDR=0.01."}
# optimization plots
result <- optimize_parameter(fun=est_power, opt1="n", opt2="lambda0", opt1Value=c(5,10,15,20),opt2Value = c(5,10,15,20), f=0.01, phi0=0.05714)
```

### Iterating and calculating powers through different sample-size (n):

```{r, power15, tidy=TRUE, warnings=FALSE}
est_power(n=8, lambda0=20, phi0=0.07154, f=0.05, m=52000,m1=71)
power10 <- est_power_distribution(n=10,f=0.05,rho=2,distributionObject=distribution,selectedGenes=genelist$Gene.name,storeProcess = TRUE)
mean(power10$power)
power13 <- est_power_distribution(n=13,f=0.05,rho=2,distributionObject=distribution,selectedGenes=genelist$Gene.name,storeProcess = TRUE)
mean(power13$power)
power15 <- est_power_distribution(n=15,f=0.05,rho=2,distributionObject=distribution,selectedGenes=genelist$Gene.name,storeProcess = TRUE)
mean(power15$power)
power18 <- est_power_distribution(n=18,f=0.05,rho=2, minAveCount=15, distributionObject=distribution,selectedGenes=genelist$Gene.name,storeProcess = TRUE)
mean(power18$power)
```


### Extract the mean gene read count across all samples:

```{r}
gene_readcounts <- distribution$pseudo.counts.mean[which(names(distribution$pseudo.counts.mean) %in% genelist$Gene.name)]
gene_dispersions <- distribution$tagwise.dispersion[which(names(distribution$pseudo.counts.mean) %in% genelist$Gene.name)]

head(gene_readcounts)

mean(gene_readcounts)
mean(gene_dispersions)

ggplot(data=data.frame(counts=as.numeric(gene_readcounts),name=as.character(genelist$Gene.name[genelist$Gene.name %in% names(gene_readcounts)]))) + geom_bar(aes(x=reorder(name,-counts),y=counts),stat="identity") + scale_y_log10() + coord_flip() + xlab("Gene name") + ylab("Mean count per gene across samples") + theme(axis.text.y = element_text(size=7)) + ggtitle("Mean count per gene across samples from experimental data") + labs(caption="Each sample has approximately 50 million reads. Experiment ID GSE58669")
```

```{r}
gene_readcounts
```